File: iEngine-Native/Engine/src/views/cameras/OrthographicCamera.cpp
#include "../../../include/iengine/views/cameras/OrthographicCamera.h"

#include <algorithm>

namespace iengine {

    OrthographicCamera::OrthographicCamera(float left, float right, float bottom, float top, float near, float far)
        : left_(left), right_(right), bottom_(bottom), top_(top), near_(near), far_(far) {
        updateProjectionMatrix();
    }

    OrthographicCamera::~OrthographicCamera() {}

    float OrthographicCamera::getLeft() const {
        return left_;
    }

    float OrthographicCamera::getRight() const {
        return right_;
    }

    float OrthographicCamera::getBottom() const {
        return bottom_;
    }

    float OrthographicCamera::getTop() const {
        return top_;
    }

    float OrthographicCamera::getNear() const {
        return near_;
    }

    float OrthographicCamera::getFar() const {
        return far_;
    }

    void OrthographicCamera::setLeft(float left) {
        left_ = left;
        projectionMatrixDirty_ = true;
    }

    void OrthographicCamera::setRight(float right) {
        right_ = right;
        projectionMatrixDirty_ = true;
    }

    void OrthographicCamera::setBottom(float bottom) {
        bottom_ = bottom;
        projectionMatrixDirty_ = true;
    }

    void OrthographicCamera::setTop(float top) {
        top_ = top;
        projectionMatrixDirty_ = true;
    }

    void OrthographicCamera::setNear(float near) {
        near_ = near;
        projectionMatrixDirty_ = true;
    }

    void OrthographicCamera::setFar(float far) {
        far_ = far;
        projectionMatrixDirty_ = true;
    }

    void OrthographicCamera::updateProjectionMatrix() {
        if (!projectionMatrixDirty_) {
            return;
        }

        // 创建正交投影矩阵
        projectionMatrix_.orthographic(left_, right_, top_, bottom_, near_, far_);

        projectionMatrixDirty_ = false;
    }

} // namespace iengine

File: iEngine-Web/src/views/cameras/OrthographicCamera.ts
import { Camera } from "./Camera";

export class OrthographicCamera extends Camera {
    // 正交参数
    public left: number = -1;
    public right: number = 1;
    public top: number = 1;
    public bottom: number = -1;
    public near: number = 0.1;
    public far: number = 100;
    
    constructor(
        left: number = -1, 
        right: number = 1, 
        top: number = 1, 
        bottom: number = -1, 
        near: number = 0.1, 
        far: number = 100
    ) {
        super();
        this.left = left;
        this.right = right;
        this.top = top;
        this.bottom = bottom;
        this.near = near;
        this.far = far;
        this.updateProjectionMatrix();
    }
    
    // 实现投影矩阵更新
    updateProjectionMatrix(): void {
        if (!this.projectionMatrixDirty) return;
        
        this.projectionMatrix.orthographic(
            this.left,
            this.right,
            this.top,
            this.bottom,
            this.near,
            this.far
        );
        
        this.projectionMatrixDirty = false;
    }
    
    // 设置正交参数
    setOrthographicParams(
        left: number, 
        right: number, 
        top: number, 
        bottom: number
    ): this {
        this.left = left;
        this.right = right;
        this.top = top;
        this.bottom = bottom;
        this.projectionMatrixDirty = true;
        return this;
    }
}